RLN: fix proof generation/verification and E2E tests

s1fr0 opened the pull request:
This WIP PR addresses https://github.com/vacp2p/zerokit/issues/16 solving previous issues with proving and verifying keys import, proof generation/verification, RLN witness fields, etc. It further removes some unnecessary crates.

oskarth commented on rln/src/circuit.rs:
```diff
@@ -33,24 +32,18 @@ pub fn VK() -> VerifyingKey<Bn254> {
-        WASM_PATH,
-        R1CS_PATH,
-    ).unwrap();  // should be )?; but need to address "the trait `From<ErrReport>` is not implemented for `protocol::ProofError`"
+    let cfg = CircomConfig::<Bn254>::new(WASM_PATH, R1CS_PATH).unwrap(); // should be )?; but need to address "the trait `From<ErrReport>` is not implemented for `protocol::ProofError`"
```
I think if you run cargo `clippy` or `fmt` it will revert this

oskarth reviewed (commented):
What specifically didn't work before but works now? Not obvious from looking at code.

Also please run `cargo clippy` and `cargo fmt` as a habit, that way diff is more clean and format is uniform.

oskarth commented on rln/src/lib.rs:
```diff
@@ -87,54 +86,64 @@ mod test {
+        let (proof, inputs) = generate_proof(builder, provingKey, rlnWitness).unwrap();
+
+        // Let's verify the proof
+        let success = verify_proof(verificationKey, proof, inputs).unwrap();
```
Why did the API change here?

s1fr0 replied:
I thought is better to standardize the APIs within each submodule in zerokit, abstracting the required inputs. For generating proof: the circom builder, a proving key and a witness. For verifying proofs the verification key, the proof and the circuit inputs to be verified.

In this way, for each submodule we need only to take care about generating witnesses, while the function signatures (modulo the type of the witness of course) will remain all the same (and do not depend on the module used). 

We can of course discuss the best approach here, not hard to change back.

oskarth replied:
OK, agree that'd be better, as long as we make the public API and C API match roughly what already exists, so should be fine

oskarth commented on rln/src/main.rs:
```diff
@@ -6,33 +6,17 @@ use ark_std::rand::thread_rng;
-
-#[derive(Debug, Deserialize)]
-//#[serde(rename_all = "camelCase")]
-struct WitnessInput {
```
Sure we don't need this anymore?

s1fr0 replied:
I moved the data structure in protocol.rs https://github.com/vacp2p/zerokit/blob/6d3571034df74e786102c06cdf4e1b78a4015094/rln/src/protocol.rs#L31 since there can be populated with some helper functions with no need to expose all fields.

s1fr0 commented:
@oskarth As we already discussed, this implementation comes from implementing from scratch RLN using semaphore primitives for Merkle tree and Poseidon hash and checking outputs at each step against zk-kit RLN. Then I integrated only the required code to make test_end_to_end pass and now I'll proceed with eliminating the unnecessary parts. 
Thus regarding the consistency of the outputs generated, I cannot guarantee if there are differences or not, since I didn't check the output of [current master](https://github.com/vacp2p/zerokit/tree/8171439a440bc48870401e4ead31ca7a48186364), since RLN was already failing.

In general the main changes/fixes were about:
- Import of ZKEYs was somehow faulty. It is interesting that with [previous import function](https://github.com/vacp2p/zerokit/blob/8171439a440bc48870401e4ead31ca7a48186364/rln/src/circuit.rs#L14) a ZKEY was correctly imported and a proof computed, but then the verification always failed, even with a good input which passes verify when the zkey is randomly populated.
- Even when a ZKEY was correctly imported with the above fix (that is that passes verify over a good input), the implemented proof function `create_proof_with_reduction_and_matrices` seemed to fail with the matrices generated from `read_zkey`. It is still unclear to me what is the reason, but I suspect that the circom builder does some backend work by itself and [`create_random_proof_with_reduction`](https://github.com/vacp2p/zerokit/blob/6d3571034df74e786102c06cdf4e1b78a4015094/rln/src/protocol.rs#L214) was enough. This made random proofs created with imported proving ZKEY correctly verified.
- The rest are utility functions to import verification keys from JSON (useful for nodes to verify proofs with no need to have a proving key), verify them against proving keys (if present), and prepare RLN witnesses from JSON or by passing each fields.
- Other minor changes;

Note that [this](https://github.com/vacp2p/zerokit/pull/18/commits/6d3571034df74e786102c06cdf4e1b78a4015094) version already has `cargo clippy; cargo fmt` applied. I'll pipe these with cargo build/run so that are applied automatically.

oskarth commented:
> Note that [this](https://github.com/vacp2p/zerokit/pull/18/commits/6d3571034df74e786102c06cdf4e1b78a4015094) version already has cargo clippy; cargo fmt applied. I'll pipe these with cargo build/run so that are applied automatically.

Oops my bad, I accidentally ended up reviewing a previous chunk.

Let me know ready for review and we can merge it

s1fr0 commented:
@oskarth ready for review. I need to think a bit more on the C APIs, so I will update them in a followup issue.

oskarth commented on rln/src/circuit.rs:
```diff
@@ -1,28 +1,134 @@
+        .into()
+}
+
+fn json_to_g1(json: &Value, key: &str) -> G1Affine {
```
Maybe a comment for this function? Not obvious what this is about

oskarth commented on rln/src/lib.rs:
```diff
@@ -25,116 +20,108 @@ mod test {
+        let root = tree.root();
+        assert_eq!(
+            root,
+            Field::from_str("0x27401a4559ce263630907ce3b77c570649e28ede22d2a7f5296839627a16e870")
```
Is this from somewhere specifically?

s1fr0 replied:
No, was randomly generated by zk-kit implementation.

oskarth commented on rln/src/protocol.rs:
```diff
@@ -103,121 +171,90 @@ pub enum ProofError {
-    proof: &Proof,
+    verifying_key: &VerifyingKey<Bn254>,
+    proof: Proof,
+    inputs: Vec<Fr>,
```
Why is this <Fr> instead of Field?

I don't quite see the purpose of removing the named parameters here as that removes context, ark_verify_proof is already general, the purpose of this function is to be specific to RLN AFAICT

s1fr0 replied:
You're right. I created a new structure which is RLNProofValues and I now pass this to verify_proof instead of inputs: Vec<Fr>. This goes into the direction to have generate_proof and verify_proof have a general API, where to the first one you pass a module-specific witness (RLNWitness) while to the second proof-values to be verified (RLNProofValues).

oskarth commented on rln/src/protocol.rs:
```diff
@@ -103,121 +171,90 @@ pub enum ProofError {
 /// # Errors
 ///
 /// Returns a [`ProofError`] if proving fails.
 pub fn generate_proof(
```
I'm still not sure I follow rationale of changing API here - the general API is `create_random_proof_with_reduction` which this specific function is hiding.

The RLNWitnessInput composite data structure seems OK to have in general, but I don't know why types are changed from Field to String etc.

Adding proving key here as requirement might be OK, depends a bit on how we imagine using this.

s1fr0 replied:
WitnessInput before had string fields https://github.com/vacp2p/zerokit/blob/master/rln/src/main.rs#L26, I just kept this. But I agree that it is better to have them another time and manage serialization on the witness data structure.

Since builder takes as inputs BigInts when pushing fields, probably this would be the best type for witness fields, so that we can avoid some conversions on the proof generation.

I agree that the general API is `create_random_proof_with_reduction`. We can either:
- preprocess builder outside and have a RLN specific proc to fill the builder and use the general API `create_random_proof_with_reduction` which takes as input builder and proving key;
- have a RLN specific `generate_proof` wrapping `create_random_proof_with_reduction` that feeds the witness into the builder and computes a proof.

In both case we will have at least 1 RLN specific procedure, unclear to me which approach is better.

oskarth commented on rln/src/protocol.rs:
```diff
@@ -103,121 +171,90 @@ pub enum ProofError {
+    mut builder: CircomBuilder<Bn254>,
+    proving_key: &ProvingKey<Bn254>,
+    rln_witness: RLNWitnessInput,
+) -> Result<(Proof, Vec<Fr>), ProofError> {
```
What is in Vec<Fr> output? inputs? aka public outputs? Can be clarified

s1fr0 replied:
Now this is removed, since users should not compute public circuit's input/output values evaluating witnesses in circuit. Utility functions will now do it directly from witnesses and generate_proof now only returns a zkSNARK proof.

oskarth commented on rln/src/public.rs:
```diff
@@ -262,26 +122,3 @@ impl Default for RLN {
 }
-
-// NOTE: Expensive test, ignoring by default
-#[ignore]
-#[test]
-fn rln_proof() {
```
This test was removed, I can't find the equivalent test that replicates this

s1fr0 replied:
You're right, is in the main now, will move.

oskarth reviewed (commented):
Good in general! See comments above

oskarth reviewed (approved):
I'm going to approve it because it simplifies integration. There are some details we might have to iterate on but this is a good start. Nice work on your first Rust PR!

oskarth commented:
Once test/lint succeeds and you are happy with it, feel free to merge. Then we can open smaller PRs for specific improvements.

s1fr0 commented:
Thanks @oskarth ! There are some other improvements implemented in this PR, which I would like to keep track of here:

- I standardized all primitives to work with semaphore's Field type, a wrapper over U256;
- Added some utility conversion from/to bytes, (bn254) Fr, and Field.
- Arithmetic over Field element is done in arkworks algebra by mapping elements to Fr and doing operation there efficiently.
- Added procedure to generate circuit's public inputs/outputs (e.g., x and y shares, nullifier, tree root, etc.) from witness. These values are stored in a new serializable struct called `RLNProofValues` which is passed to `verify_proof`.

These changes go towards https://github.com/vacp2p/zerokit/issues/19.
